### PR TITLE
kube-scheduler: NewFramework function to pass the context parameter

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
follow up https://github.com/kubernetes/kubernetes/pull/116842#discussion_r1179556544

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by alculquicondor** [comment]
is this still a draft?

**Comment by mengjiao-liu** [comment]
> is this still a draft?

No, PR has finished and the test passed. I forgot to update my status. Updated.

File: pkg/scheduler/framework/autoscaler_contract/framework_contract_test.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ func TestFrameworkContract(t *testing.T) {
 
 func TestNewFramework(t *testing.T) {
 	var f interface{}
-	if f, _ = runtime.NewFramework(nil, nil, nil); f != nil {
+	if f, _ = runtime.NewFramework(context.Background(), nil, nil); f != nil {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/autoscaler_contract/framework_contract_test.go:46]: build a context from `t`

File: pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go [hunk 2]
```diff
@@ -108,7 +108,7 @@ func TestRequestedToCapacityRatioScoringStrategy(t *testing.T) {
 
 			state := framework.NewCycleState()
 			snapshot := cache.NewSnapshot(test.existingPods, test.nodes)
-			fh, _ := runtime.NewFramework(nil, nil, ctx.Done(), runtime.WithSnapshotSharedLister(snapshot))
+			fh, _ := runtime.NewFramework(ctx, nil, nil, runtime.WithSnapshotSharedLister(snapshot))
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go:111]: also should build a context from `t`

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go [hunk 3]
```diff
@@ -2377,7 +2377,7 @@ func BenchmarkFilter(b *testing.B) {
 		b.Run(tt.name, func(b *testing.B) {
 			existingPods, allNodes, _ := st.MakeNodesAndPodsForEvenPodsSpread(tt.pod.Labels, tt.existingPodsNum, tt.allNodesNum, tt.filteredNodesNum)
 			ctx := context.Background()
-			pl := plugintesting.SetupPlugin(b, topologySpreadFunc, &config.PodTopologySpreadArgs{DefaultingType: config.ListDefaulting}, cache.NewSnapshot(existingPods, allNodes))
+			pl := plugintesting.SetupPlugin(ctx, b, topologySpreadFunc, &config.PodTopologySpreadArgs{DefaultingType: config.ListDefaulting}, cache.NewSnapshot(existingPods, allNodes))
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:2380]: here too

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go [hunk 4]
```diff
@@ -1366,7 +1366,7 @@ func BenchmarkTestPodTopologySpreadScore(b *testing.B) {
 		b.Run(tt.name, func(b *testing.B) {
 			existingPods, allNodes, filteredNodes := st.MakeNodesAndPodsForEvenPodsSpread(tt.pod.Labels, tt.existingPodsNum, tt.allNodesNum, tt.filteredNodesNum)
 			state := framework.NewCycleState()
-			pl := plugintesting.SetupPlugin(b, podTopologySpreadFunc, &config.PodTopologySpreadArgs{DefaultingType: config.ListDefaulting}, cache.NewSnapshot(existingPods, allNodes))
+			pl := plugintesting.SetupPlugin(context.Background(), b, podTopologySpreadFunc, &config.PodTopologySpreadArgs{DefaultingType: config.ListDefaulting}, cache.NewSnapshot(existingPods, allNodes))
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go:1369]: build from `b`

File: pkg/scheduler/framework/plugins/selectorspread/selector_spread_test.go [hunk 5]
```diff
@@ -403,7 +402,7 @@ func TestSelectorSpreadScore(t *testing.T) {
 			if err != nil {
 				t.Errorf("error creating informerFactory: %+v", err)
 			}
-			fh, err := frameworkruntime.NewFramework(nil, nil, ctx.Done(), frameworkruntime.WithSnapshotSharedLister(snapshot), frameworkruntime.WithInformerFactory(informerFactory))
+			fh, err := frameworkruntime.NewFramework(ctx, nil, nil, frameworkruntime.WithSnapshotSharedLister(snapshot), frameworkruntime.WithInformerFactory(informerFactory))
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/selectorspread/selector_spread_test.go:405]: here too

File: pkg/scheduler/framework/plugins/selectorspread/selector_spread_test.go [hunk 6]
```diff
@@ -661,7 +660,7 @@ func TestZoneSelectorSpreadPriority(t *testing.T) {
 			if err != nil {
 				t.Errorf("error creating informerFactory: %+v", err)
 			}
-			fh, err := frameworkruntime.NewFramework(nil, nil, wait.NeverStop, frameworkruntime.WithSnapshotSharedLister(snapshot), frameworkruntime.WithInformerFactory(informerFactory))
+			fh, err := frameworkruntime.NewFramework(ctx, nil, nil, frameworkruntime.WithSnapshotSharedLister(snapshot), frameworkruntime.WithInformerFactory(informerFactory))
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/selectorspread/selector_spread_test.go:663]: same

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go [hunk 7]
```diff
@@ -789,7 +788,7 @@ func TestVolumeBinding(t *testing.T) {
 				runtime.WithClientSet(client),
 				runtime.WithInformerFactory(informerFactory),
 			}
-			fh, err := runtime.NewFramework(nil, nil, wait.NeverStop, opts...)
+			fh, err := runtime.NewFramework(ctx, nil, nil, opts...)
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go:791]: you might have solved a leaked routine with this :)

create the context from `t`.

File: pkg/scheduler/framework/preemption/preemption_test.go [hunk 8]
```diff
@@ -284,7 +284,8 @@ func TestDryRunPreemption(t *testing.T) {
 			ctx, cancel := context.WithCancel(context.Background())
```
Comment (alculquicondor) [inline pkg/scheduler/framework/preemption/preemption_test.go:284]:
```suggestion
			ctx, cancel := context.WithCancel(ctx)
```

The one coming from `NewTestContext`

File: pkg/scheduler/framework/runtime/framework.go [hunk 9]
```diff
@@ -250,13 +250,14 @@ func defaultFrameworkOptions(stopCh <-chan struct{}) frameworkOptions {
 var _ framework.Framework = &frameworkImpl{}
 
 // NewFramework initializes plugins given the configuration and the registry.
-func NewFramework(r Registry, profile *config.KubeSchedulerProfile, stopCh <-chan struct{}, opts ...Option) (framework.Framework, error) {
+func NewFramework(ctx context.Context, r Registry, profile *config.KubeSchedulerProfile, opts ...Option) (framework.Framework, error) {
+	stopCh := ctx.Done()
 	options := defaultFrameworkOptions(stopCh)
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:255]:
```suggestion
	options := defaultFrameworkOptions(ctx.Done())
```

File: pkg/scheduler/framework/runtime/framework_test.go [hunk 10]
```diff
@@ -514,11 +515,12 @@ func TestNewFrameworkErrors(t *testing.T) {
 
 	for _, tc := range tests {
 		t.Run(tc.name, func(t *testing.T) {
+			_, ctx := ktesting.NewTestContext(t)
```
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework_test.go:518]: add a cancel

File: pkg/scheduler/scheduler.go [hunk 11]
```diff
@@ -239,14 +239,14 @@ var defaultSchedulerOptions = schedulerOptions{
...
+	client clientset.Interface,
 	informerFactory informers.SharedInformerFactory,
 	dynInformerFactory dynamicinformer.DynamicSharedInformerFactory,
 	recorderFactory profile.RecorderFactory,
-	stopCh <-chan struct{},
 	opts ...Option) (*Scheduler, error) {
 
-	stopEverything := stopCh
+	stopEverything := ctx.Done()
 	if stopEverything == nil {
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:250]: do we still need this check?

File: pkg/scheduler/scheduler_test.go [hunk 12]
```diff
@@ -173,14 +174,14 @@ func TestSchedulerCreation(t *testing.T) {
 
 			eventBroadcaster := events.NewBroadcaster(&events.EventSinkImpl{Interface: client.EventsV1()})
 
-			stopCh := make(chan struct{})
-			defer close(stopCh)
+			ctx, cancel := context.WithCancel(context.Background())
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler_test.go:177]: create a context from `t`

File: pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go (hunk 3 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/plugins/podtopologyspread/filtering_test.go:2380]: Done.

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go (hunk 4 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go:1371]: Done.

File: pkg/scheduler/framework/plugins/selectorspread/selector_spread_test.go (hunk 5 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/plugins/selectorspread/selector_spread_test.go:407]: Done.

File: pkg/scheduler/framework/plugins/selectorspread/selector_spread_test.go (hunk 6 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/plugins/selectorspread/selector_spread_test.go:666]: Done.

File: pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go (hunk 7 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go:793]: Done.

File: pkg/scheduler/framework/runtime/framework_test.go (hunk 10 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/runtime/framework_test.go:518]: Done.

File: pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go (hunk 2 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/plugins/noderesources/requested_to_capacity_ratio_test.go:112]: Done.

File: pkg/scheduler/scheduler.go (hunk 11 above)
Comment (mengjiao-liu) [reply pkg/scheduler/scheduler.go]: Looks like we won't need this check. It has been removed.

**Comment by mengjiao-liu** [comment]
All comments addressed. Squash commits after tests passed.

File: pkg/scheduler/framework/runtime/framework.go (hunk 9 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/runtime/framework.go]: Done.

File: pkg/scheduler/framework/preemption/preemption_test.go (hunk 8 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/preemption/preemption_test.go]: Updated.

File: pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go [hunk 13]
```diff
@@ -1364,12 +1366,13 @@ func BenchmarkTestPodTopologySpreadScore(b *testing.B) {
 	}
 	for _, tt := range tests {
 		b.Run(tt.name, func(b *testing.B) {
+			_, ctx := ktesting.NewTestContext(b)
 			existingPods, allNodes, filteredNodes := st.MakeNodesAndPodsForEvenPodsSpread(tt.pod.Labels, tt.existingPodsNum, tt.allNodesNum, tt.filteredNodesNum)
 			state := framework.NewCycleState()
-			pl := plugintesting.SetupPlugin(b, podTopologySpreadFunc, &config.PodTopologySpreadArgs{DefaultingType: config.ListDefaulting}, cache.NewSnapshot(existingPods, allNodes))
+			pl := plugintesting.SetupPlugin(context.Background(), b, podTopologySpreadFunc, &config.PodTopologySpreadArgs{DefaultingType: config.ListDefaulting}, cache.NewSnapshot(existingPods, allNodes))
```
Comment (alculquicondor) [inline pkg/scheduler/framework/plugins/podtopologyspread/scoring_test.go:1372]:
```suggestion
			pl := plugintesting.SetupPlugin(ctx, b, podTopologySpreadFunc, &config.PodTopologySpreadArgs{DefaultingType: config.ListDefaulting}, cache.NewSnapshot(existingPods, allNodes))
```

File: pkg/scheduler/scheduler.go [hunk 14]
```diff
@@ -283,9 +280,9 @@ func New(client clientset.Interface,
 
 	snapshot := internalcache.NewEmptySnapshot()
 	clusterEventMap := make(map[framework.ClusterEvent]sets.Set[string])
-	metricsRecorder := metrics.NewMetricsAsyncRecorder(1000, time.Second, stopCh)
+	metricsRecorder := metrics.NewMetricsAsyncRecorder(1000, time.Second, stopEverything)
```
Comment (alculquicondor) [inline pkg/scheduler/scheduler.go:283]:
```suggestion
	metricsRecorder := metrics.NewMetricsAsyncRecorder(1000, time.Second, ctx.Done())
```

**Review by alculquicondor — COMMENTED** [review]
just a couple of misses

/approve

File: pkg/scheduler/framework/autoscaler_contract/framework_contract_test.go [hunk 15]
```diff
@@ -42,8 +43,9 @@ func TestFrameworkContract(t *testing.T) {
 }
 
 func TestNewFramework(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
 	var f interface{}
-	if f, _ = runtime.NewFramework(nil, nil, nil); f != nil {
+	if f, _ = runtime.NewFramework(ctx, nil, nil); f != nil {
```
Comment (alculquicondor) [inline pkg/scheduler/framework/autoscaler_contract/framework_contract_test.go:48]: /assign @x13n 
for review from CA.

File: pkg/scheduler/scheduler.go (hunk 14 above)
Comment (mengjiao-liu) [reply pkg/scheduler/scheduler.go:283]: I think it's equivalent. 
```
stopEverything := ctx.Done() 
```

The `stopEverything` variable is used in other places as well(Line 326,330,339).
We don't need to call `ctx.Done()` again.

**Comment by mengjiao-liu** [comment]
Modified to fix conflict because PR #117834 was merged first.

file: `pkg/scheduler/schedule_one_test.go:529`

**Comment by mengjiao-liu** [comment]
pull-kubernetes-kind-dra CI job  has been failing since 5-21 

issue: https://github.com/kubernetes/kubernetes/issues/118201

**Comment by mengjiao-liu** [comment]
The issue https://github.com/kubernetes/kubernetes/issues/118201 has been solved.
/test pull-kubernetes-kind-dra

**Comment by pohly** [comment]
/approve

For test/integration.

@alculquicondor: will you do the final LGTM?

**Review by alculquicondor — COMMENTED** [review]
/lgtm

**Pull request merged**